### PR TITLE
End to end SQL flow test

### PR DESCRIPTION
**Pull request by achalpandey**
I also removed InputFlowletConfigurationTest.

### CODE REVIEW

**Comment by gokulavasan** [comment]
@achalpandey Does this PR cover announcing of endpoints through FlowletContext in the ready callback method?

**Comment by achalpandey** [comment]
No thats not in this PR. I only made all the changes required for running the end to end test. 

As discussed offline I'll create a new PR for it.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+public class SQLFlowTest extends TestBase {
+  private static FlowManager flowManager;
+  private static Thread ingestData;
+  static final int MAX_TIMESTAMP = 10;
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    Map<String, String> runtimeArgs = Maps.newHashMap();
+    //Identifying free port
+    ServerSocket serverSocket = new ServerSocket(0);
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: Use a function for this. Also the code needs to close socket in finally.

``` java
 public static int getRandomPort() {
    try {
      ServerSocket socket = new ServerSocket(0);
      try {
        return socket.getLocalPort();
      } finally {
        socket.close();
      }
    } catch (IOException e) {
      return -1;
    }
  }
```

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    final int port = serverSocket.getLocalPort();
+    serverSocket.close();
+    serverSocket = new ServerSocket(0);
+    int tcpPort = serverSocket.getLocalPort();
+    serverSocket.close();
+    runtimeArgs.put(Constants.HTTP_PORT, Integer.toString(port));
+    runtimeArgs.put(Constants.TCP_PORT + "_intInput", Integer.toString(tcpPort));
+    flowManager = deployFlow(SQLFlow.class, runtimeArgs);
+    TimeUnit.SECONDS.sleep(15);
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: Why have random sleep? Can we make this robust by having a smaller sleep and checking for status until a timeout.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            JsonArray dataArray = new JsonArray();
+            dataArray.add(new JsonPrimitive(i));
+            dataArray.add(new JsonPrimitive(j));
+            bodyJson.add("data", dataArray);
+            HttpPost httpPost = new HttpPost("http://localhost:" + port + "/v1/tigon/intInput");
+            StringEntity params = null;
+            try {
+              params = new StringEntity(bodyJson.toString());
+            } catch (UnsupportedEncodingException e) {
+              e.printStackTrace();
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: Let's not catch the exception the unit test should fail in that case, which will happen if exception is raised

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        HttpClient httpClient = new DefaultHttpClient();
+        for (int i = 1; i <= MAX_TIMESTAMP; i++) {
+          for (int j = 1; j <= i; j++) {
+            JsonObject bodyJson = new JsonObject();
+            JsonArray dataArray = new JsonArray();
+            dataArray.add(new JsonPrimitive(i));
+            dataArray.add(new JsonPrimitive(j));
+            bodyJson.add("data", dataArray);
+            HttpPost httpPost = new HttpPost("http://localhost:" + port + "/v1/tigon/intInput");
+            StringEntity params = null;
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: no need to initialize null

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      public void run() {
+        HttpClient httpClient = new DefaultHttpClient();
+        for (int i = 1; i <= MAX_TIMESTAMP; i++) {
+          for (int j = 1; j <= i; j++) {
+            JsonObject bodyJson = new JsonObject();
+            JsonArray dataArray = new JsonArray();
+            dataArray.add(new JsonPrimitive(i));
+            dataArray.add(new JsonPrimitive(j));
+            bodyJson.add("data", dataArray);
+            HttpPost httpPost = new HttpPost("http://localhost:" + port + "/v1/tigon/intInput");
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: We can just use a simple UrlConnection and not have dependency on org.apache. See CDAP code base for examples.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            StringEntity params = null;
+            try {
+              params = new StringEntity(bodyJson.toString());
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: Always use Charsets UTF-8.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  class DataPacket {
+    //Using the same data type and variable name as specified in the query output
+    long timestamp;
+    int sumValue;
+
+    public String toString() {
+      return "timestamp : " + timestamp + "\tsumValue : " + sumValue;
+    }
+  }
+}
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: new line

File: tigon-sql/src/main/java/co/cask/tigon/sql/conf/Constants.java [hunk 8]
```diff
@@ -59,4 +59,10 @@
    */
   public static final String QTREE = "qtree.xml";
   public static final String OUTPUT_SPEC = "output_spec.cfg";
+
+  /**
+   * Ports Map Keys
+   */
+  public static final String HTTP_PORT = "httpPort";
+  public static final String TCP_PORT = "tcpPort";
```
Comment (sreevatsanraman) [inline tigon-sql/src/main/java/co/cask/tigon/sql/conf/Constants.java]: Call this TCP_INGESTION_PORT_PREFIX. and add "_" in the end.

File: tigon-sql/src/main/java/co/cask/tigon/sql/flowlet/AbstractInputFlowlet.java [hunk 9]
```diff
@@ -130,6 +131,22 @@ public void initialize(FlowletContext ctx) throws Exception {
     create(configurer);
     InputFlowletSpecification spec = configurer.createInputFlowletSpec();
 
+    Map<String, Integer> dataIngestionPortsMap = Maps.newHashMap();
+    if (ctx.getRuntimeArguments().get(Constants.HTTP_PORT) != null) {
+      // Get data ingestion ports from runtime args
+      dataIngestionPortsMap.put(Constants.HTTP_PORT, Integer.parseInt(ctx.getRuntimeArguments().get(Constants.HTTP_PORT)));
+      for (String inputName : spec.getInputSchemas().keySet()) {
+        dataIngestionPortsMap.put(Constants.TCP_PORT + "_" + inputName,
+                    Integer.parseInt(ctx.getRuntimeArguments().get(Constants.TCP_PORT + "_" + inputName)));
```
Comment (sreevatsanraman) [inline tigon-sql/src/main/java/co/cask/tigon/sql/flowlet/AbstractInputFlowlet.java]: What happens if users want random HTTP port and specify TCP port configuration in runtime args. These checks need to be independent.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java (hunk 2 above)
Comment (achalpandey) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: This is the same issue as on the integration test PR https://github.com/caskco/tigon/pull/13

Unfortunately for now we don't have a better way to address this issue. We'll have to address this in a subsequent release.

**Comment by achalpandey** [comment]
@sreevatsanraman I addressed your comments. Please review.

File: tigon-sql/src/main/java/co/cask/tigon/sql/ioserver/StreamEngineIO.java [hunk 10]
```diff
@@ -48,11 +48,13 @@
   private final Map<String, InetSocketAddress> dataSourceServerMap = Maps.newHashMap();
   private final GDATRecordQueue recordQueue;
   private DataIngestionRouter router;
+  private final Map<String, Integer> portMap;
 
   //TODO Remove GDATRecordQueue parameter from this constructor. Use Guice to inject it directly to OutputServerSocket
```
Comment (sreevatsanraman) [inline tigon-sql/src/main/java/co/cask/tigon/sql/ioserver/StreamEngineIO.java:53]: Create OSS jira for TODO and add it in code.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    runtimeArgs.put(Constants.TCP_INGESTION_PORT_PREFIX + "intInput", Integer.toString(tcpPort));
+    flowManager = deployFlow(SQLFlow.class, runtimeArgs);
+    TimeUnit.SECONDS.sleep(30);
+    ingestData = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          for (int i = 1; i <= MAX_TIMESTAMP; i++) {
+            for (int j = 1; j <= i; j++) {
+              HttpURLConnection urlConn =
```
Comment (sreevatsanraman) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java]: Close URL connection in finally block

**Comment by sreevatsanraman** [comment]
lgtm :+1:

**Comment by achalpandey** [comment]
Bamboo https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT12

**Comment by achalpandey** [comment]
Thanks for the comments! :) I fixed things as per your suggestion. Will merge after the bamboo build completes.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+  }
+
+  private static DataPacket getDataPacket() throws IOException {
+    HttpClient httpClient = new DefaultHttpClient();
+    HttpGet httpGet = new HttpGet(serviceURL + "/poll-data");
+    HttpResponse response = httpClient.execute(httpGet);
+    if (response.getStatusLine().getStatusCode() != 200) {
+      return null;
+    }
+    String[] requestData = EntityUtils.toString(response.getEntity(), Charsets.UTF_8).replace("\"", "").split(":");
```
Comment (ParthGandhi) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java:141]: This is overly complicated. Simply use GSON. 

To serialize in the handler:

```
Gson gson = new Gson();
String json = gson.toJson(dataPacket)
```

To deserialize here:

```
DataPacket dataPacket = gson.fromJson(responseString, DataPacket.class);  
```

https://sites.google.com/site/gson/gson-user-guide#TOC-Serializing-and-Deserializing-Collection-with-Objects-of-Arbitrary-Types

**Comment by ParthGandhi** [comment]
File a JIRA for my comment. 

LGTM.

File: tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java (hunk 12 above)
Comment (achalpandey) [inline tigon-sql/src/test/java/co/cask/tigon/sql/sqlflow/SQLFlowTest.java:141]: https://issues.cask.co/browse/TIGON-6

**Comment by achalpandey** [comment]
Bamboo GREEN

**Pull request merged**